Overhauled debugging memory leaks (#573)
apple/swift-org-website: documentation/server/guides/memory-leaks-and-usage.md

Debugging memory leaks in Swift on both macOS and Linux environments can be done using different tools and techniques, each with distinct strengths and usability. 

Basic troubleshooting steps include:

1. Using profiling tools, provided by the respective operating systems and development environments, to identify and analyze memory usage.

    **For macOS**, [Memory Graph Debugger](https://developer.apple.com/documentation/xcode/gathering-information-about-memory-use#Inspect-the-debug-memory-graph) and this [Detect and diagnose memory issues](https://developer.apple.com/videos/play/wwdc2021/10180/) video are helpful. You can also use the [Xcode Instruments](https://help.apple.com/instruments/mac/10.0/#/dev022f987b) tool for various profiling instruments including the [Allocations instrument](https://developer.apple.com/documentation/xcode/gathering-information-about-memory-use#Profile-your-app-using-the-Allocations-instrument) to track memory allocation and deallocation in your Swift code. 
    

>     **For server Linux**, you can use tools like [Valgrind](https://valgrind.org/) or [Heaptrack](https://github.com/KDE/heaptrack) to profile your application as shown in the examples below. Although these tools are primarily used for C/C++ code, they can also work with Swift.

Reviewer: This seems redundant

```suggestion
    **For Linux**, you can use tools like [Valgrind](https://valgrind.org/) or [Heaptrack](https://github.com/KDE/heaptrack) to profile your application as shown in the examples below. Although these tools are primarily used for C/C++ code, they can also work with Swift.
```

Author: Ack

> Note: Memory leaks can occur in Swift if there are substantial reference cycles between objects that involve closures or if objects hold references to external resources that are not released properly. However, the likelihood of such issues is significantly reduced through the automatic memory management's ability to add and remove references, making sources of leaks like retained references and unbalanced retain-release cycles less common in Swift code.

3. Enabling debug memory allocation features allows you to get additional information about objects and their memory allocations.
    
    **On macOS**, you can enable Zombie Objects using Xcode or use [MallocStackLogging](https://developer.apple.com/videos/play/wwdc2022/10106/) to detect over-released or accessed deallocated objects.
    
    To enable Zombie Objects: 
    1. Open your Xcode project.
    2. Go to the **Edit Scheme** menu by clicking on the scheme dropdown in the toolbar. 
    3. In the scheme editor window, select the **Run** tab.
    4. Choose the **Diagnostics** tab.
    5. Under **Memory Management**, check the box next to **Enable Zombie Objects**.
    

>     **On server Linux**, Swift has built-in LeakSanitizer support that can be enabled using the `-sanitize=leak` compiler flag.

Reviewer: As above

```suggestion
    **On Linux**, Swift has built-in LeakSanitizer support that can be enabled using the `-sanitize=leak` compiler flag.
```

Author: ack

```
public class MemoryLeaker {
   var closure: () -> Void = { () }
   
   public init() {}
   
   public func doNothing() {}
   
   public func doSomethingThatLeaks() {
      self.closure = {
         // This will leak as it'll create a permanent reference cycle:
         //
         //     self -> self.closure -> self
         self.doNothing()
      }
   }
}
@inline(never) // just to be sure to get this in a stack trace
func myFunctionDoingTheAllocation() {
   let thing = MemoryLeaker()
   thing.doSomethingThatLeaks()
}

myFunctionDoingTheAllocation()
```
## Debugging leaks with Valgrind
Valgrind is an open-source framework for debugging and profiling Linux applications. It provides several tools, including Memcheck, which can detect memory leaks, invalid memory accesses, and other memory errors. Although Valgrind is primarily focused on C/C++ applications, it can also be used with Swift on Linux.

To debug memory leaks using Valgrind, install it on your system. 

For MacOS:
1. Open a Terminal session.
2. [Install Homebrew](https://brew.sh/) if you haven't already. 
3. Once `Homebrew` is installed, run this command to install `valgrind`:

> ```

Reviewer: I think these blocks can be marked as bash
```suggestion
```bash
```

Author: Ack

brew install valgrind
```

4. Enter your password if prompted to authorize the software installation and allow Homebrew to complete the installation process. Confirm the installation when asked.

Valgrind should be successfully installed on your system using the system package manager.

5. Run this `valgrind` command to enable full leak checking:

```

> valgrind --leak-check=full ./test

Reviewer: It's probably good to at least mention that `./test` refers to the compiled program and if we need to pass any flags through to the compiler etc

Author: ack. this is from the existing documentation, can you please clarify what you're thinking and/or provide some language or an example?

Reviewer: Something like

> once you've compiled your program, in this case a binary named `test`, run this `valgrind` command...

Author: Ack. Updated language. Thanks

==1==    by 0x108E58: $s4test12MemoryLeakerCACycfC (in /tmp/test)
==1==    by 0x10900E: $s4test28myFunctionDoingTheAllocationyyF (in /tmp/test)
==1==    by 0x108CA3: main (in /tmp/test)

```

However, since Swift uses name mangling for function and symbol names, the stack traces may not be straightforward to understand. 

To demangle the Swift symbols in the stack traces, run the `swift demangle` command:

```
swift demangle <mangled_symbol>
```

> Replace `<mangled_symbol>` with the mangled symbol name shown in the stack trace. 

Reviewer: It would be good to show an example of exactly what a mangled symbol is in case people get confused with addresses etc
```suggestion
Replace `<mangled_symbol>` with the mangled symbol name shown in the stack trace, e.g. `swift demangle $s4test12MemoryLeakerCACycfC`
```

Author: ack

> Replace `<mangled_symbol>` with the mangled symbol name shown in the stack trace. For example, `<swift demangle $s4test12MemoryLeakerCACycfC>`

Reviewer: ```suggestion
Replace `<mangled_symbol>` with the mangled symbol name shown in the stack trace. For example, `swift demangle $s4test12MemoryLeakerCACycfC`
```

    #4 0x7f7e43aecb96  (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)

SUMMARY: AddressSanitizer: 32 byte(s) leaked in 1 allocation(s).

```

> Unfortunately, the output doesn’t provide a human-readable representation of the function names because [LeakSanitizer doesn't symbolicate stack traces on Linux](https://bugs.swift.org/browse/SR-12601). 

Reviewer: Worth adding a "yet"?

Also, we need to update the URL to point to GH now that bugs.swift.org is no longer a thing

Author: Which URL?

Reviewer: ```suggestion
Unfortunately, the output doesn’t provide a human-readable representation of the function names because [LeakSanitizer doesn't symbolicate stack traces on Linux](https://github.com/apple/swift/issues/55046). 
```

Author: Updated language and link.

However, you can symbolicate it using `llvm-symbolizer` or `addr2line` if you have `binutils` installed.

To install `binutils` for Swift on a server running Linux, follow these steps:
1. Connect to your Swift server through SSH using a terminal.
2. Update the package lists by running the following command:

```
sudo apt update
```

3. Install `binutils` by running the following command:

```
sudo apt install binutils
```

> 4.  This will install `binutils` and its related tools for working with binaries, object files, and libraries, which can be useful for developing and debugging Swift applications on a Linux server.

Reviewer: ```suggestion
4.  This will install `binutils` and its related tools for working with binaries, object files, and libraries, which can be useful for developing and debugging Swift applications on Linux.
```

Author: ack

* LeakSanitizer is primarily supported on macOS and Linux. While it is possible to use LeakSanitizer on iOS or other platforms that support Swift, there may be limitations or platform-specific issues that need to be considered.
* Enabling Address Sanitizer and LeakSanitizer in your Swift project can have a performance impact. It is recommended to use LeakSanitizer for targeted analysis and debugging rather than continuously running it in production environments.

## Debugging transient memory usage with Heaptrack
[Heaptrack](https://github.com/KDE/heaptrack) is an open-source heap memory profiler tool that is helpful for finding and analyzing memory leaks and usage with less overhead than Valgrind. It also allows for analyzing and debugging transient memory usage in your application. However, it may significantly impact performance by overloading the allocator.

A GUI front-end analyzer `heaptrack_gui` is available in addition to command line access. The analyzer allows for diffing between two different runs of your application to troubleshoot variations in `malloc` behavior between the `feature branch` and `main`.

> Using a different example, here’s a short how-to using [Ubuntu 20.04](https://www.swift.org/download/) to analyze transient usage.

Reviewer: We should bump this to the latest Swift we support as 20.04 only has a year of support left, maybe just change to Ubuntu and lose the link

Author: ack.

Heaptrack finished! Now run the following to investigate the data:

  heaptrack --analyze "/tmp/.nio_alloc_counter_tests_GRusAy/heaptrack.test_1000_autoReadGetAndSet.84341.gz"

```

> 3. Then run it a second time for the `feature branch`:

Reviewer: We should add a mention that we need to change the branch and recompile

Author: Noted. will revise.

      at /home/ubuntu/swiftnio/swift-nio/Sources/NIO/LinuxURing.swift:297
...
22196 temporary allocations of 22276 allocations in total (99.64%) from:

```

Looking at the output above, we can see the extra transient allocations were due to extra debug printing and querying of environment variables as shown below:

```
NIO.URing.getEnvironmentVar(Swift.String) -> Swift.String?
  at /home/ubuntu/swiftnio/swift-nio/Sources/NIO/LinuxURing.swift:291
  in /tmp/.nio_alloc_counter_tests_GRusAy/.build/x86_64-unknown-linux-gnu/release/test_1000_autoReadGetAndSet
NIO.URing._debugPrint(@autoclosure () -> Swift.String) -> ()

```

> Since the code will be removed before final integration of the `feature branch`, the diff will also disappear.

Reviewer: This sentence is a bit confusing without code to refer to and makes a number of assumptions. Either we need to show the code or say something like "in this example the debug prints are only for testing and would be removed from the code before the branch is merged"

Author: noted. will revise.
